src/librustc_mir/transform/check_consts/validation.rs

```diff
@@ -0,0 +1,597 @@
+                    Some(def) if def.is_union() => {
+                        self.check_op(ops::UnionAccess);
+                    }
+
+                    _ => {}
+                }
+            }
+
+            ProjectionElem::Downcast(..) => {
+                self.check_op(ops::Downcast);
+            }
+        }
+    }
+
+
+    fn visit_source_info(&mut self, source_info: &SourceInfo) {
+        trace!("visit_source_info: source_info={:?}", source_info);
+        self.span = source_info.span;
+    }
+
+    fn visit_statement(&mut self, statement: &Statement<'tcx>, location: Location) {
+        trace!("visit_statement: statement={:?} location={:?}", statement, location);
+
+        self.qualifs.needs_drop.visit_statement(statement, location);
+        self.qualifs.has_mut_interior.visit_statement(statement, location);
+        debug!("needs_drop: {:?}", self.qualifs.needs_drop.get());
+        debug!("has_mut_interior: {:?}", self.qualifs.has_mut_interior.get());
+
+        match statement.kind {
+            StatementKind::Assign(..) => {
+                self.super_statement(statement, location);
+            }
+            StatementKind::FakeRead(FakeReadCause::ForMatchedPlace, _) => {
+                self.check_op(ops::IfOrMatch);
+            }
+            // FIXME(eddyb) should these really do nothing?
+            StatementKind::FakeRead(..) |
+            StatementKind::SetDiscriminant { .. } |
+            StatementKind::StorageLive(_) |
+            StatementKind::StorageDead(_) |
+            StatementKind::InlineAsm {..} |
+            StatementKind::Retag { .. } |
+            StatementKind::AscribeUserType(..) |
+            StatementKind::Nop => {}
+        }
+    }
+
+    fn visit_terminator(&mut self, terminator: &Terminator<'tcx>, location: Location) {
+        trace!("visit_terminator: terminator={:?} location={:?}", terminator, location);
+
+        self.qualifs.needs_drop.visit_terminator(terminator, location);
+        self.qualifs.has_mut_interior.visit_terminator(terminator, location);
+        debug!("needs_drop: {:?}", self.qualifs.needs_drop.get());
+        debug!("has_mut_interior: {:?}", self.qualifs.has_mut_interior.get());
+
+        self.super_terminator(terminator, location);
+    }
+
+    fn visit_terminator_kind(&mut self, kind: &TerminatorKind<'tcx>, location: Location) {
+        trace!("visit_terminator_kind: kind={:?} location={:?}", kind, location);
+        self.super_terminator_kind(kind, location);
+
+        match kind {
+            TerminatorKind::Call { func, .. } => {
+                let fn_ty = func.ty(self.body, self.tcx);
+
+                let def_id = match fn_ty.kind {
+                    ty::FnDef(def_id, _) => def_id,
+
+                    ty::FnPtr(_) => {
+                        self.check_op(ops::FnCallIndirect);
+                        return;
+                    }
+                    _ => {
+                        self.check_op(ops::FnCallOther);
```

Comment:
Callees must always be `FnPtr` or `FnDef` (Miri will ICE when it sees anything else), how is this here even reachable?

Reply:
This const-checking pass has structured errors everywhere the old pass had a call to `span_err`. If you would like to change this, it would be nice to encode the set of types that can appear as the `func` operand of a `Call` terminator in an `enum` somewhere so we could use exhaustive matching in more places.

Reply:
>  it would be nice to encode the set of types that can appear as the func operand of a Call terminator in an enum somewhere so we could use exhaustive matching in more places.

This is basically suggesting "intrinsically-typed MIR", where invalid MIR cannot even be constructed. I am not sure what the best way would be to do that in Rust (I only know how to do it with dependent types). Calls are not the only operations this affects, tons of MIR operations have invariants for which types can occur -- arithmetic primitives, casts, pointer dereferences, ...

Reply:
I just meant an `as_func_ty` combinator for `Ty` whose return value is restricted to the types that are valid as a `func`.

Reply:
I just meant that there are more "subclasses" of types than "callable types", and some of them overlap. I don't think we should adopt something ad-hoc that only works for callable types.

Reply:
Gotta start somewhere. What do we do for closures btw? Do these get lowered to `FnDecl`s by this point?

Reply:
> Gotta start somewhere.

I'd rather just remove this dead code for now, and leave improving MIR itself for later. Do you want to track this in an issue?

Reply:
Closures do not get called directly. Instead `FnOnce::call_once` (or the other `Fn*` traits) gets called, which is a `FnDef`. The `ty::Closure` is then passed as `self` argument.

See for example [the MIR of this function](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8fe8ef9f867615cf2c9deab88bfeeac2).